lapack/gonum/dgesc2.go

```diff
@@ -0,0 +1,71 @@
+// Copyright ©2021 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package gonum
+
+import (
+	"math"
+
+	"gonum.org/v1/gonum/blas/blas64"
+)
+
+// Dgesc2 solves a system of linear equations
+//  A * X = scale * RHS
+// with a general N-by-N matrix A using the LU factorization with
+// complete pivoting computed by Dgetc2. The result is placed in
+// rhs on exit.
+//
+// Dgesc2 is an internal routine. It is exported for testing purposes.
+func (impl Implementation) Dgesc2(n int, a []float64, lda int, rhs []float64, ipiv, jpiv []int) (scale float64) {
+	switch {
```

Comment:
This part should look like this:
```
	switch {
	case n < 0:
		panic(nLT0)
	case lda < max(1, n):
		panic(badLdA)
    }

    // Quick return if possible.
    if n == 0 {
        return 0
    }

    switch {
	case len(a) < (n-1)*lda+n:
		panic(shortA)
	case len(rhs) < n:
		panic(shortRHS)
	case len(ipiv) != n:
	    panic(badLenIpiv)
	case len(jpiv) != n:
	    panic(badLenJpiv)
	}

	const smlnum = dlamchS / dlamchP
```

Reply:
I've applied your suggestions. I'm still confused as to how much we have to test. Is it enough that it did not panic with a bunch of random matrices? Should the result be compared with another linear solver?

Reply:
We must test that, given the output from Dgetc2, Dgesc2 computes what it promises, that is, that:
```
A * X = scale * RHS
```
holds and that `scale` is in `[0,1]`.

I would start with:
1. Generate a random A, compute the norm |A|
2. Generate a random X, compute the norm |X|
3. Compute RHS from A*X
4. Call Dgetc2
5. Call Dgesc2
6. Check that scale is in [0,1]
7. Check that |A*X - scale*RHS| / |A| / |X| is small

If nothing special is done, the above steps will always return `scale = 1`. Then the next step would be to modify 1. and/or 2. to trigger a non-unit scaling.